pkg/attribute/mutableBag.go

```diff
@@ -15,330 +15,347 @@
+			}
 
-	// my kingdom for a clear method on maps!
+			delete(mb.values, name)
+		}
+	}
 
-	for k := range mb.strings {
-		delete(mb.strings, k)
+	// apply all attributes
+	for k, v := range attrs.StringAttributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating string attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.int64s {
-		delete(mb.int64s, k)
+	for k, v := range attrs.Int64Attributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating int64 attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.float64s {
-		delete(mb.float64s, k)
+	for k, v := range attrs.DoubleAttributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating double attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.bools {
-		delete(mb.bools, k)
+	for k, v := range attrs.BoolAttributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating bool attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.times {
-		delete(mb.times, k)
+	for k, v := range attrs.TimestampAttributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating time attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.durations {
-		delete(mb.durations, k)
+	for k, v := range attrs.DurationAttributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating duration attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.bytes {
-		delete(mb.bytes, k)
+	for k, v := range attrs.BytesAttributes {
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating bytes attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
+		}
+		mb.values[dictionary[k]] = v
 	}
 
-	for k := range mb.stringMaps {
-		delete(mb.stringMaps, k)
+	for k, v := range attrs.StringMapAttributes {
+		m, ok := mb.values[dictionary[k]].(map[string]string)
+		if !ok {
+			m = make(map[string]string)
+			mb.values[dictionary[k]] = m
+		}
+
+		if log != nil {
+			log.WriteString(fmt.Sprintf("  updating stringmap attribute %s from\n", dictionary[k]))
+
+			if len(m) > 0 {
+				for k2, v2 := range m {
+					log.WriteString(fmt.Sprintf("    %s:%s\n", k2, v2))
+				}
+			} else {
+				log.WriteString("    <empty>\n")
+			}
+
+			log.WriteString("  to\n")
+		}
+
+		for k2, v2 := range v.Map {
+			m[dictionary[k2]] = v2
+		}
+
+		if log != nil {
+			if len(m) > 0 {
+				for k2, v2 := range m {
+					log.WriteString(fmt.Sprintf("    %s:%s\n", k2, v2))
+				}
+			} else {
+				log.WriteString("    <empty>\n")
```

Comment:
aside: all the "\n"s in this file make me wonder if we should wrap the log.WriteString in some sort of helper method that always appends a new line...